cmd/internal/database/meilisearch/meilisearch.go

```diff
@@ -0,0 +1,329 @@
+	dump := path.Join(constants.RestoreDir, latestStableDump)
+	if _, err := os.Stat(dump); os.IsNotExist(err) {
+		return fmt.Errorf("restore file not present: %s", dump)
+	}
+	start := time.Now()
+
+	if err := utils.RemoveContents(db.dbdir); err != nil {
+		return fmt.Errorf("could not clean database data directory: %w", err)
+	}
+
+	err := db.importDump(ctx, dump)
+	if err != nil {
+		return fmt.Errorf("unable to recover %w", err)
+	}
+
+	db.log.Infow("recovery done", "duration", time.Since(start))
+	return nil
+}
+
+// Upgrade implements database.Database.
+func (db *Meilisearch) Upgrade(ctx context.Context) error {
+	start := time.Now()
+
+	versionFile := path.Join(db.dbdir, meilisearchVersionFile)
+	if _, err := os.Stat(versionFile); errors.Is(err, fs.ErrNotExist) {
+		db.log.Infof("%q is not present, no upgrade required", versionFile)
+		return nil
+	}
+
+	dbVersion, err := db.getDatabaseVersion(versionFile)
+	if err != nil {
+		return err
+	}
+	meilisearchVersion, err := db.getBinaryVersion(ctx)
+	if err != nil {
+		return err
+	}
+	if (dbVersion.Major() == meilisearchVersion.Major()) && (dbVersion.Minor() == meilisearchVersion.Minor()) {
+		db.log.Infow("no version difference, no upgrade required", "database-version", dbVersion, "binary-version", meilisearchVersion)
+		return nil
+	}
+	if dbVersion.GreaterThan(meilisearchVersion) {
+		db.log.Errorw("database is newer than meilisearch binary, aborting", "database-version", dbVersion, "binary-version", meilisearchVersion)
+		return fmt.Errorf("database is newer than meilisearch binary")
+	}
+	if _, err := os.Stat(db.latestStableDumpDst); errors.Is(err, fs.ErrNotExist) {
+		return fmt.Errorf("%q is not present, no upgrade possible, maybe no backup was running before", db.latestStableDumpDst)
+	}
+
+	db.log.Infow("start upgrade", "from", dbVersion, "to", meilisearchVersion)
+
+	err = os.Rename(db.dbdir, db.dbdir+".old")
+	if err != nil {
+		return fmt.Errorf("unable to rename dbdir: %w", err)
+	}
+
+	err = db.importDump(ctx, db.latestStableDumpDst)
+	if err != nil {
+		return err
+	}
+	db.log.Infow("upgrade done and new data in place", "took", time.Since(start))
+	return nil
+}
+
+func (db *Meilisearch) importDump(ctx context.Context, dump string) error {
+	var (
+		err  error
+		cmd  *exec.Cmd
+		g, _ = errgroup.WithContext(ctx)
+	)
+
+	g.Go(func() error {
+		args := []string{"--import-dump", dump, "--master-key", db.apikey}
+		db.log.Infow("execute meilisearch", "args", args)
+
+		cmd = exec.CommandContext(ctx, meilisearchCmd, args...) // nolint:gosec
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+		err = cmd.Run()
+		if err != nil {
+			return fmt.Errorf("unable import dump %w", err)
+		}
+		db.log.Info("import of dump finished")
+		return nil
+	})
+
+	// TODO big databases might take longer, not sure if 100 attempts are enough
+	// must check how long it take max with backoff ?
+	err = retry.Do(func() error {
+		v, err := db.client.Version()
```

Comment:
Would be nice to reuse `Probe` here and also add the health `isHealthy` check to `Probe` as it probably makes sense to only start with taking backups when the database is healthy.